Remove problematic circular reference

geichelberger opened the pull request:
This PR contains the following changes:

- Made additional ACL actions organization-specific and configurable via organization properties.
- Removed the circular dependency between the Elasticsearch index and ListProvidersService
- Moved resource list providers into a shared module, split list-provider service into API/impl, and cleaned up the namespace.
- Removed ACL additional actions list provider config in favor of the organization properties-based implementation.

Fixes #4117

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

geichelberger committed:
Add organization-specific ACL additional actions

Enable defining organization-specific additional actions that can be
used in ACLs. The implementation reads organization properties with the
prefix org.opencastproject.acl.additional.actions. and returns an
unmodifiable Map of actionName -> value, allowing ACL rules to be
extended per organization without changing global configuration.

geichelberger committed:
Remove circular dependency ListProvidersService

Break the circular dependency by removing the direct dependency from
ElasticsearchIndex to ListProvidersService. Some list providers depend
on index service, which created a problematic cycle. This change
decouples the index implementation from the list-provider lookup,
reducing coupling and avoiding initialization/order problems.

geichelberger commented:
@gregorydlogan I could try to prepare a version that would not require a configuration change and use the current config as a fallback, so that it could be part of 19.x. Or should it still go into develop? This could reduce the occurrences of circular reference exceptions.

gregorydlogan commented:
It would be nice to have this in 19, so yeah if you can make this work without the org config modifications that would be awesome!

mtneug commented:
What is the status of this? There has been a case mentioned in Matrix recently.

@geichelberger do I read your comment correctly that there could be a (partial) fix for `19.x` and presumably a full fix for `develop`?

Arnei commented:
Code looks reasonable to me, so I'd be glad if we could move forward with this.

geichelberger commented:
The fix wouldn't be partial; it's just a backward-compatibility layer to support the old configuration file from the list provider, which I would remove in another PR from the development version.

geichelberger committed:
Refactor resource list provider namespace

This extracts all resource list providers from the index service bundle
and moves them to a common list provider bundle. Further, the
list-providers-service module is split up into an api and an impl module.

geichelberger committed:
Remove acl additional action listprovider config

This replaces the listprovider config with a organization config based
implementation.

geichelberger commented:
During the dev meeting, the decision was made to target the develop branch. I extended the approach a bit by extracting the list-providers from the index-service bundle into a separate list-providers bundle.

geichelberger committed:
Remove authorization-manager dependency

Remove the dependency to the authorization-manager from the
list-providers-common bundle.

geichelberger committed:
Move agents listprovider

geichelberger committed:
Move theme and event comments listprovider

Arnei reviewed (approved):
Code looks reasonable to me. Spun up a local distributed setup and tested some ACL stuff. Also restarted the admin somewhere between 10 and 20 times, did not see a circular reference error (which is of course no guarantee at all that there are none left, but it is still nice).

geichelberger committed:
Fix RolesListProvider package

geichelberger commented:
I backported it to 19.x, and we have been using it in production since today.

gregorydlogan commented:
This is ready to go @geichelberger?

gregorydlogan commented on modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java:
```diff
@@ -59,6 +62,8 @@ private SecurityUtil() {
   /** The name of the key used to store the name of the system user in the global config. */
   public static final String PROPERTY_KEY_SYS_USER = "org.opencastproject.security.digest.user";
 
+  public static final String PROPERTY_KEY_ACL_ADDITIONL_ACTIONS = "org.opencastproject.acl.additional.actions.";
```
Are you trying to keep under a variable length rule, or is this just a typo? :)

```suggestion
  public static final String PROPERTY_KEY_ACL_ADDITIONAL_ACTIONS = "org.opencastproject.acl.additional.actions.";
```

gregorydlogan reviewed (changes requested):
Needs a release notes entry as well referencing this PR.

Otherwise looks good and seems to work correctly.  I haven't seen the race, but then given that we're actively removing how it happens that's kinda expected :)

geichelberger committed:
Fix typo in property constant

geichelberger committed:
Add release note

geichelberger commented:
This should be ready to go :)
